### ISSUE TITLE
Python 3.9: multiple bases have instance lay-out conflict with persistent 4.5.1

### ISSUE THREAD

**Issue by tschorr** [zopefoundation/Persistence#22]
Running tests with Python 3.9-dev result in the following traceback:

```python traceback
Traceback (most recent call last):
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/__main__.py", line 19, in <module>
    run()
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/__init__.py", line 30, in run
    failed = run_internal(defaults, args, script_parts=script_parts, cwd=cwd,
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/__init__.py", line 51, in run_internal
    runner.run()
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/runner.py", line 179, in run
    feature.global_setup()
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/find.py", line 518, in global_setup
    tests = find_tests(self.runner.options, self.runner.found_suites)
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/find.py", line 173, in find_tests
    for suite in found_suites:
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/find.py", line 194, in find_suites
    for fpath, package in find_test_files(options):
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/find.py", line 260, in find_test_files
    for f, package in find_test_files_(options):
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/find.py", line 288, in find_test_files_
    for (p, package) in test_dirs(options, {}):
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/find.py", line 343, in test_dirs
    p = import_name(p)
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/zope/testrunner/find.py", line 417, in import_name
    __import__(name)
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/Persistence/__init__.py", line 60, in <module>
    from Persistence.mapping import PersistentMapping  # NOQA
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/Persistence/mapping.py", line 81, in <module>
    class PersistentMapping(with_metaclass(
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/Persistence/mapping.py", line 74, in __new__
    return meta(name, resolved_bases, d)
  File "/home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/Persistence/mapping.py", line 32, in __new__
    return ExtensionClass.__new__(cls, *args, **kw)
TypeError: multiple bases have instance lay-out conflict
```

This error is raised from https://github.com/python/cpython/blob/9aeb0ef9309384099e2f23bcee2240fbc096568e/Objects/typeobject.c#L2029 because none of the provided base classes is a `PyType_IsSubtype` of the other. Verify by looking at the bases' mro:

```pycon
> /home/thomas/devel/plone/Persistence/.tox/py39/lib/python3.9/site-packages/Persistence/mapping.py(74)__new__()
-> return meta(name, resolved_bases, d)
(Pdb) p resolved_bases
(<class 'Persistence.Persistent'>, 
  <class 'persistent.mapping.PersistentMapping'>)
(Pdb) p resolved_bases[0].__mro__
(<class 'Persistence.Persistent'>, 
  <class 'persistent.Persistent'>, 
  <class 'ExtensionClass.Base'>, 
  <class 'object'>)
(Pdb) p resolved_bases[1].__mro__
(<class 'persistent.mapping.PersistentMapping'>, 
  <class 'collections.UserDict'>, 
  <class 'collections.abc.MutableMapping'>, 
  <class 'collections.abc.Mapping'>, 
  <class 'collections.abc.Collection'>, 
  <class 'collections.abc.Sized'>, 
  <class 'collections.abc.Iterable'>, 
  <class 'collections.abc.Container'>, 
  <class 'persistent.persistence.Persistent'>, 
  <class 'object'>)
```

The mro looks different in Python 3.8.1 where the tests are passing:

```pycon
> /home/thomas/devel/plone/Persistence/.tox/py38/lib/python3.8/site-packages/Persistence/mapping.py(74)__new__()
-> return meta(name, resolved_bases, d)
(Pdb) p resolved_bases
(<class 'Persistence.Persistent'>, 
  <class 'persistent.mapping.PersistentMapping'>)
(Pdb) p resolved_bases[0].__mro__
(<class 'Persistence.Persistent'>, 
  <class 'persistent.Persistent'>, 
  <class 'ExtensionClass.Base'>, 
  <class 'object'>)
(Pdb) p resolved_bases[1].__mro__
(<class 'persistent.mapping.PersistentMapping'>, 
  <class 'collections.UserDict'>, 
  <class 'collections.abc.MutableMapping'>, 
  <class 'collections.abc.Mapping'>, 
  <class 'collections.abc.Collection'>, 
  <class 'collections.abc.Sized'>, 
  <class 'collections.abc.Iterable'>, 
  <class 'collections.abc.Container'>, 
  <class 'persistent.Persistent'>, 
  <class 'object'>)
```

-> here  `resolved_bases[1]` includes the C implementation of `persistent.Persistent` in the mro, as opposed to the Python implementation `persistent.persistence.Persistent` in Python 3.9.

This is caused by an import error in `persistent.__init__.py`, s. https://github.com/zopefoundation/persistent/issues/124

**Comment by tschorr** [comment]
seen on Ubuntu bionic.

**Comment by icemac** [comment]
zopefoundation/persistent#124 is fixed now and released in 4.6.1.

I saw AccessControl tests switching to green when using the new `persistent` version.
Can this issue be closed or is there still something to do?

**Comment by icemac** [comment]
I think this issue was resolved by zopefoundation/persistent#124.

**Comment by icemac** [comment]
Feel free to reopen if this is not the case.
